feat(cohorts): Recalculate all affected cohorts on create, update

dustinbyrne opened the pull request:
## Problem

Currently when a cohort is recalculated, it also recalculates its dependent graph. E.g., in the case of:
```
A <- B <- C
```

If C updates, so will B, then A. However the reverse is not true. If A updates, B and C will still reference stale data until they're recalculated by schedule or by manual intervention.

Resolves #36134

## Changes

When recalculating a cohort, its dependency graph is now also automatically enqueued for recalculation. This change uses the new [cohort dependency cache](#37897) to build the dependency graph.

## How did you test this code?

Automated tests are added to validate:
- Recalculation properly enqueues dependencies for recalulation
- API integration tests to ensure recalculation works as expected when changing or creating cohorts
- Model properties for cohort dependents / dependencies

dustinbyrne committed:
feat(cohorts): handle recalculation for all related cohorts (dependencies and dependents)

dustinbyrne committed:
test(cohorts): add tests for enqueueing cohort recalculations with dependencies and dependents

dustinbyrne committed:
test(cohorts): add API test for recalculation across cohort dependency chains

gustavohstrassburger reviewed (approved):
Nice work.
I left two comments, but they are non-blockers.

dustinbyrne committed:
refactor: `get_cohort_dependents` accepts int id

This allows callers to continuously retrieve the dependent chain without
hitting the database - assuming the cache is warm.

dustinbyrne committed:
fix: `get_all_cohort_dependents` queries once

This change drops the `seen_cohorts_cache` argument in favor of querying
once to retrieve all cohorts. In practice, this should be much faster
considering `seen_cohorts_cache` wasn't used anywhere.

dustinbyrne commented on posthog/models/cohort/util.py:
```diff
@@ -556,6 +557,52 @@ def get_all_cohort_dependencies(
+        seen_cohorts_cache = {}
+
+    cohorts = []
+    seen_cohort_ids = {cohort.id}
```
Thanks for pointing this out! It is confusing, and I'd forgotten about this. `seen_cohorts_cache` was a pattern used by the similar method `get_all_cohort_dependencies`. Long story short, it's used in that method but never used by callers for this new method. I've removed it, and pushed the query into the caching strategy. In practice now, this method will only ever execute a single db query when resolving dependent IDs to Cohorts.
